File: problems/0015-3sum/solution.go
// 15. 3Sum https://leetcode.com/problems/3sum/
package main

import (
	"fmt"
	"slices"
)

// threeSum 은 합이 0인 중복 없는 삼중항을 모두 반환한다.
func threeSum(nums []int) [][]int {
	// TODO: 여기에 풀이를 작성하세요.
	slices.Sort(nums)
	var res [][]int
	for i := 0; i < len(nums)-2; i++ {
		num := nums[i]
		if num > 0 { // 더 이상은 불가
			break
		}
		if i > 0 && num == nums[i-1] { // 중복 제거
			continue
		}

		st := i + 1
		end := len(nums) - 1
		for st < end {
			sum := num + nums[st] + nums[end]
			if sum == 0 {
				res = append(res, []int{num, nums[st], nums[end]})
				st++
				for st < end && nums[st] == nums[st-1] { // 중복 제거
					st++
				}
			} else if sum < 0 {
				st++
			} else {
				end--
			}
		}

	}

	return res
}

// main 은 손으로 굴려보는 놀이터. `go run ./problems/0015-3sum/`
func main() {
	fmt.Println(threeSum([]int{-1, 0, 1, 2, -1, -4})) // [[-1 -1 2] [-1 0 1]]
	fmt.Println(threeSum([]int{0, 1, 1}))             // []
	fmt.Println(threeSum([]int{0, 0, 0}))             // [[0 0 0]]
}

File: problems/0015-3sum/solution.py
"""15. 3Sum — https://leetcode.com/problems/3sum/"""


def three_sum(nums: list[int]) -> list[list[int]]:
    """합이 0인 중복 없는 삼중항을 모두 반환한다.

    시간 O(n^2) — 정렬 O(n log n) + 바깥 루프 n * 투 포인터 O(n)
    공간 O(1) — 정렬이 제자리, 출력 제외
    """
    nums.sort()
    res = []
    n = len(nums)

    for i in range(n - 2):
        # 정렬돼 있으므로 고정값이 양수면 뒤는 전부 양수 → 합 0 불가
        if nums[i] > 0:
            break
        # 고정값이 직전과 같으면 결과가 통째로 중복된다
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        # 남은 구간에서 합이 -nums[i]인 두 수 찾기 (Two Sum)
        l, r = i + 1, n - 1
        while l < r:
            s = nums[i] + nums[l] + nums[r]
            if s < 0:
                # 가장 큰 짝과 붙여도 모자라니 nums[l]은 가망 없음
                l += 1
            elif s > 0:
                r -= 1
            else:
                res.append([nums[i], nums[l], nums[r]])
                l += 1
                # 같은 l 값은 같은 삼중항을 또 만든다
                while l < r and nums[l] == nums[l - 1]:
                    l += 1
    return res


if __name__ == "__main__":
    print(three_sum([-1, 0, 1, 2, -1, -4]))  # [[-1, -1, 2], [-1, 0, 1]]
